### PR TITLE
Fixed terminator deserialisation, and fixed encoding type.

### PR DESCRIPTION
**Pull request by CitadelCore**
One problem I found is that the binary deserialiser does not read the terminators at the end of the document in a stream. This is fine when you have one file with plain binary keyvalues1, but when you have a file with multiple sections with custom headers (like `appinfo.vdf`), you run into problems when reading the next section's header because there are still terminator bytes that haven't been read.

Additionally, keys are saved as ASCII and values as UTF-8. If the file contains UTF-8 data (like the multilingual content in `appinfo.vdf`) there is the potential for data loss.

I've implemented a fix for both of these problems - the former, I've modified `KV1BinarySerializer.cs` to read all terminator bytes at the end, and the latter, I've modified both `KV1BinarySerializer.cs` and `KV1BinaryReader.cs` to read UTF-8 rather than ASCII in object string values.

All unit tests pass. Hope these improvements will prove useful.

### CODE REVIEW

File: ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs [hunk 1]
```diff
@@ -17,19 +17,19 @@ public KV1BinaryReader(Stream stream, IVisitationListener listener)
...
+            _stream = stream;
+            _listener = listener;
+            _reader = new BinaryReader(stream);
         }
 
-        readonly Stream stream;
-        readonly BinaryReader reader;
-        readonly IVisitationListener listener;
-        bool disposed;
+        private readonly Stream _stream;
```
Comment (xPaw) [inline ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: Can you revert these changes please?

It makes for an unnecessarily hard diff to read.

Comment (CitadelCore) [reply ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: Done

File: ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs [hunk 2]
```diff
@@ -61,82 +61,81 @@ public void Dispose()
...
-                    throw new ArgumentOutOfRangeException(nameof(type));
+                    listener.OnKeyValuePair(name, value);
+                }
             }
-
-            listener.OnKeyValuePair(name, value);
         }
 
-        string ReadNullTerminatedString()
+        string ReadNullTerminatedString(Encoding encoding = null)
```
Comment (yaakov-h) [inline ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: This seems misleading to me because it only works with encoding that are terminated by a single NUL byte. I'm not 100% sure this actually covers UTF-8 correctly, but it definitely doesn't work with UTF-16 or UTF-32. Perhaps this function should just read the null terminated byte stream, and the caller should decode.

File: ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs [hunk 3]
```diff
@@ -61,82 +61,81 @@ public void Dispose()
...
+            if (encoding == null) encoding = Encoding.ASCII;
+            var b = reader.ReadByte();
 
-            return sb.ToString();
+            int len;
+            for (len = 0; b != 0; len++)
+                b = reader.ReadByte();
+
+            reader.BaseStream.Position -= len + 1;
+            var str = reader.ReadBytes(len);
```
Comment (yaakov-h) [inline ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: We don't need to read twice. Consider keeping a buffer, and resizing it if we overrun.

File: ValveKeyValue/ValveKeyValue/Serialization/KV1BinarySerializer.cs [hunk 4]
```diff
@@ -78,13 +78,12 @@ void Write(KV1BinaryNodeType nodeType)
...
         }
 
-        void Write(string value)
+        void Write(string value, Encoding encoding = null)
         {
-            foreach (var @char in value)
-            {
-                writer.Write(@char);
-            }
+            if (encoding == null) encoding = Encoding.ASCII;
```
Comment (yaakov-h) [inline ValveKeyValue/ValveKeyValue/Serialization/KV1BinarySerializer.cs]: As above regarding NUL termination, but we don't really have a better option here.

File: ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs [hunk 5]
```diff
@@ -61,82 +61,81 @@ public void Dispose()
...
-            var name = ReadNullTerminatedString();
-            ReadValue(name, type);
-        }
-
-        void ReadValue(string name, KV1BinaryNodeType type)
-        {
-            KVValue value;
-
-            switch (type)
+            while (true)
```
Comment (yaakov-h) [inline ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: I need to check the larger structure, but looping here seems wrong. We're trying to read one object, not many. If we're just not reading `End` bytes then we should add that, as a peek if necessary.

Comment (CitadelCore) [reply ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: I've removed the loop and moved logic back to ReadObjectCore and ReadValue, which works fine.

File: ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs (hunk 3 above)
Comment (CitadelCore) [reply ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: Now using MemoryStream.

File: ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs (hunk 2 above)
Comment (CitadelCore) [reply ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: Moved to a function that just reads the null-terminated byte array, and moved the actual decoding to the callers.

File: ValveKeyValue/ValveKeyValue/Serialization/KV1BinarySerializer.cs (hunk 4 above)
Comment (CitadelCore) [reply ValveKeyValue/ValveKeyValue/Serialization/KV1BinarySerializer.cs]: Modified to write a terminated byte array instead of a string and moved encoding to the caller.

**Comment by CitadelCore** [comment]
As well as the changes in the comments, I've added UTF-8 tests with Chinese characters to `BinaryObjectSerializationTestCase` and `BinaryObjectDeserializationTestCase.` I've also added two new unit tests, `BinaryObjectConsecutiveSerializationTestCase` and `BinaryObjectConsecutiveDeserializationTestCase`, for consecutive serialisation and deserialisation.

I've added the class `NonClosingStream` as a test helper, since the deserialiser will automatically close the stream when it finishes, which is not what we want when consecutively reading objects. Perhaps there should be an option to leave the stream open like the one in `BinaryWriter`, which would avoid this issue entirely.

File: ValveKeyValue/ValveKeyValue.Test/Helpers/NonClosingStream.cs [hunk 6]
```diff
@@ -0,0 +1,311 @@
+﻿using System;
+using System.IO;
+
+namespace CSERLibrary.Utilities
```
[the remark this answers is not in the document]
Comment (CitadelCore) [reply ValveKeyValue/ValveKeyValue.Test/Helpers/NonClosingStream.cs]: Woops - I copied it from my own project, which in itself obtained it from [www.yoda.archsys.com/csharp/miscutil/](url). I've fixed the namespace. The original code is under the Apache license.

[the remark this answers is not in the document]
Comment (xPaw) [reply ValveKeyValue/ValveKeyValue.Test/Helpers/NonClosingStream.cs]: `NonClosingStream` will not be needed if you add `leaveOpen: true` where it is missing as @yaakov-h suggested instead.

[the remark this answers is not in the document]
Comment (CitadelCore) [reply ValveKeyValue/ValveKeyValue.Test/Helpers/NonClosingStream.cs]: Fixed. KV1BinaryReader both didn't have `leaveOpen: true` in the BinaryReader constructor and additionally disposed the stream manually in its Dispose method.

File: ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs [hunk 7]
```diff
@@ -54,79 +54,79 @@ public void Dispose()
...
-                case KV1BinaryNodeType.Int64:
-                    value = new KVObjectValue<long>(reader.ReadInt64(), KVValueType.Int64);
-                    break;
-
-                default:
-                    throw new ArgumentOutOfRangeException(nameof(type));
+                listener.OnObjectStart(name);
+                ReadObjectCore();
+                listener.OnObjectEnd();
+            }
```
Comment (xPaw) [inline ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: If you `return` in the if, then you won't need to nest everything in the `else`.

File: ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs [hunk 8]
```diff
@@ -54,79 +54,79 @@ public void Dispose()
...
         void ReadObjectCore()
         {
-            var type = ReadNextNodeType();
-            var name = ReadNullTerminatedString();
-            ReadValue(name, type);
+            KV1BinaryNodeType type;
+
+            // Keep reading values, until we reach the terminator
+            while ((type = ReadNextNodeType()) != KV1BinaryNodeType.End)
+                ReadValue(type);
```
Comment (xPaw) [inline ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: Add braces please (and in other places too)

File: ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs [hunk 9]
```diff
@@ -54,79 +54,79 @@ public void Dispose()
...
         }
 
-        void ReadValue(string name, KV1BinaryNodeType type)
+        void ReadValue(KV1BinaryNodeType type)
         {
-            KVValue value;
-
-            switch (type)
+            // ASCII encoding is used for keys
```
[the remark this answers is not in the document]
Comment (CitadelCore) [reply ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: I just did some testing and actually, there seems to be no problems using UTF8 for the keys,

File: ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs (hunk 7 above)
Comment (CitadelCore) [reply ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: Don't see why this can't be done? You've struck it through.

Comment (CitadelCore) [reply ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: Implemented @yaakov-h's suggestion.

File: ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (xPaw) [reply ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: @yaakov-h what do you think about this?

[the remark this answers is not in the document]
Comment (CitadelCore) [reply ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: Changed to UTF8.

File: ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs (hunk 8 above)
Comment (CitadelCore) [reply ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: Changed.

File: ValveKeyValue/ValveKeyValue.Test/Binary/BinaryObjectConsecutiveDeserializationTestCase.cs [hunk 10]
```diff
@@ -0,0 +1,68 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
```
Comment (xPaw) [inline ValveKeyValue/ValveKeyValue.Test/Binary/BinaryObjectConsecutiveDeserializationTestCase.cs]: What is this used by? From a quick glance I can't see it, but I haven't fired up VS to check.

Comment (CitadelCore) [reply ValveKeyValue/ValveKeyValue.Test/Binary/BinaryObjectConsecutiveDeserializationTestCase.cs]: A garbage using that was automatically inserted by resharper when I copied over some of my own code. Removed it.

Comment (xPaw) [reply ValveKeyValue/ValveKeyValue.Test/Binary/BinaryObjectConsecutiveDeserializationTestCase.cs]: Ah.

Comment (yaakov-h) [reply ValveKeyValue/ValveKeyValue.Test/Binary/BinaryObjectConsecutiveDeserializationTestCase.cs]: It’s still there, though?

File: ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (yaakov-h) [reply ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs]: @xPaw if only there was an actual spec...
